linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -4683,4 +4683,191 @@ private void flush() {
       }
     };
   }
+
+  /**
+   * Merge Union Enumerable.
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   *
+   * @param sources input enumerables (must be already sorted)
+   * @param sortKeySelector sort key selector
+   * @param sortComparator sort comparator to decide the next item
+   * @param all whether duplicates will be considered or not
+   * @param equalityComparer {@link EqualityComparer} to control duplicates,
+   *                         only used if {@code all} is {@code false}
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  public static <TSource, @Nullable TKey> Enumerable<TSource> mergeUnion(
+      List<Enumerable<TSource>> sources,
+      Function1<TSource, @Nullable TKey> sortKeySelector,
```

Comment:
It would be slightly better to have something like `Function1<? super TSource, ? extends TKey>` so users could pass method references with subtypes. I don't know if that combines with other types, however, in most cases `Function1` arguments should look like `Map#computeIfAbsent`:

```java
public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
```

Reply:
Everywhere in EnumerableDefaults we have `Function1<TSource, TKey>` I think we should keep it like this for consistency.
Maybe your proposal could be addressed as separate evolution to be performed in bulk in all EnumerableDefaults methods. with this type of function pattern.